Add openpmd-pipe.py command line tool

franzpoeschel opened the pull request:
This adds a little Python-based command line tool that can pipe data from a source to a sink.
Also adds handling of attributes by explicit datatype to the Python API to make generic attribute writing safe.

Possible uses:
* Converting openPMD-based data from one backend to another one.
* Multiplexing in staging setups

### TODO
- [x] Documentation
- [x] Cleanup
- [x] Testing: Resulting files must again be convertible using `openpmd-pipe`
- [x] Integration into the build system
- [x] Switch to a serial mode if MPI is not configured
- [x] Particle Patches
- [x] rebase after #912 was merged

ax3l commented on src/cli/pipe.py:
```diff
@@ -0,0 +1,243 @@
+This module provides functions that are wrapped into sys.exit(...()) calls by
+the setuptools (setup.py) "entry_points" -> "console_scripts" generator.
+
+Copyright 2020 openPMD contributors
```
```suggestion
Copyright 2021 openPMD contributors
```

ax3l commented:
Looks good!
Do you potentially want to implement this as a python module so people can use it with `python -m openpmd.pipe ...` and then add a console_script hook on top? See `openpmd-ls` for an example.

The reason behind this is that not all distributions add paths like `~/.local/bin` to `PATH` and thus some `pip` users will not be able to use it as a CLI tool but need to go via the module call.

franzpoeschel commented:
> Looks good!
> Do you potentially want to implement this as a python module so people can use it with `python -m openpmd.pipe ...` and then add a console_script hook on top? See `openpmd-ls` for an example.
> 
> The reason behind this is that not all distributions add paths like `~/.local/bin` to `PATH` and thus some `pip` users will not be able to use it as a CLI tool but need to go via the module call.

Can you check whether I've done this somewhat correctly?

franzpoeschel commented:
I've rebased this on to #912 to remove the GC workaround. Otherwise, this is ready for review from my side :)

ax3l commented:
Now that #912 is merged, this PR needs a little rebase against `dev` :)
(Don't forget to remove the `[WIP]` to signal it's ready for review & merge :) )

ax3l commented on src/binding/python/Attributable.cpp:
```diff
@@ -289,6 +317,18 @@ void init_Attributable(py::module &m) {
+                Attributable & attr,
+                std::string const& key,
+                py::object& obj,
+                std::string const& datatype )
```
I usually try to use the `np.dtype` for all datatype options in the Python bindings

ax3l commented on src/binding/python/Attributable.cpp:
```diff
@@ -335,6 +375,10 @@ void init_Attributable(py::module &m) {
             return v.getResource();
             // TODO instead of returning lists, return all arrays (ndim > 0) as numpy arrays?
         })
+        .def("get_attribute_type", []( Attributable & attr, std::string const& key ) {
```
Ideally we would also return the `numpy.dtype` here for consistency.
```suggestion
        .def("get_attribute_dtype", []( Attributable & attr, std::string const& key ) {
```

ax3l replied:
oh, and we can make this a (read-only) `property` such as `attribute_dtype`

franzpoeschel replied:
Wouldn't a property only work for something without parameters?

Recent commit uses numpy types. There are some workarounds involved, but it's working as long as noone gets the idea to define an attribute of type `array<double,7>` not named `unitDimension`.

I also took the freedom to add a version of `switchType` that only switches over non-vector types, I was annoyed of always having to define a hundred of extra structs. Also, I made two improvements to that version: (1) not having to give the return type any more and (2) making the `action.template operator()<0>(...)` (resp. 1000) optional.

ax3l replied:
Oh, right - didn't see that here.

Probably a cleaner interface on the long term, as in h5py, would be `<an attributable>.attrs` as a key-value store and then we can ask the `.dtype` of the numpy scalars/arrays inside.

Switchtype updates are great, I have used this way too little in the Python bindings yet.

franzpoeschel replied:
That would be a breaking interface change though, deprecating our current `Attributable.attributes` call, right?
For now, I've turned this into a readonly property `Attributable.attribute_dtypes` (returning a dictionary: name -> numpy type).

franzpoeschel committed:
Python Attributable: Write attributes by explicity datatype

Adds Attributable.get_attribute_type that returns a string-based type
identifier for a given attribute.
A further Overload of Attributable.set_attribute allows explicitly
specifying a type using such an identifier.

franzpoeschel committed:
Add openpmd_api.pipe module and executable

This includes install scripts and tests in CMake

franzpoeschel committed:
Use numpy types

ax3l commented on include/openPMD/Datatype.hpp:
```diff
@@ -772,6 +771,182 @@ ReturnType switchType( Datatype dt, Action action, Args &&... args )
+            std::forward< Args >( args )... );
+    case Datatype::DATATYPE:
+        return detail::CallUndefinedDatatype<
+            1000, ReturnType, Action, void, Args &&... >::
```
Can we express that magic number in other way?
(Last entry of the data type enum + 1, or a constexpr with a docstring, or so?)

Maybe the same for the `0` below, just to make things more clear by naming them (first/last/lowest/highest-something).

franzpoeschel replied:
I did not yet apply that to the original switchType yet since I'm planning a separate PR for that one.

ax3l commented on include/openPMD/Datatype.hpp:
```diff
@@ -772,6 +771,182 @@ ReturnType switchType( Datatype dt, Action action, Args &&... args )
+ * @tparam n As in switchType().
+ * @tparam ReturnType As in switchType().
+ * @tparam Action As in switchType().
+ * @tparam Dummy For SFINAE, set to void.
```
For [inclusive wording](https://developers.google.com/style/inclusive-documentation) we could replace _dummy_ with _placeholder_.

ax3l commented on CMakeLists.txt:
```diff
@@ -881,6 +884,16 @@ if(openPMD_INSTALL)
+                install(
+                    FILES ${openPMD_SOURCE_DIR}/src/cli/${toolname}.py
+                    DESTINATION ${CMAKE_INSTALL_BINDIR}
+                    RENAME openpmd-${toolname}.py
```
I would call the tool just `openpmd-pipe` because for the user it will make no difference if the tool is a script or an executable:
```suggestion
                    RENAME openpmd-${toolname}
```

ax3l commented on CMakeLists.txt:
```diff
@@ -1088,19 +1116,74 @@ if(openPMD_BUILD_TESTING)
+            add_custom_command(TARGET openPMD.py.cliTools POST_BUILD
+                COMMAND ${CMAKE_COMMAND} -E copy
+                        ${openPMD_SOURCE_DIR}/src/cli/${toolname}.py
+                        ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/openpmd-${toolname}.py
```
```suggestion
                        ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/openpmd-${toolname}
```

ax3l commented on CMakeLists.txt:
```diff
@@ -1088,19 +1116,74 @@ if(openPMD_BUILD_TESTING)
+            )
+            add_test(NAME CLI.help.${toolname}.py
+                COMMAND ${Python_EXECUTABLE}
+                    ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/openpmd-${toolname}.py --help
```
```suggestion
                    ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/openpmd-${toolname} --help
```

ax3l commented on CMakeLists.txt:
```diff
@@ -1088,19 +1116,74 @@ if(openPMD_BUILD_TESTING)
+                add_test(NAME CLI.pipe.py
+                    COMMAND sh -c
+                        "${MPI_TEST_EXE} ${Python_EXECUTABLE}                      \
+                            ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/openpmd-pipe.py      \
```
```suggestion
                            ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/openpmd-pipe         \
```

ax3l commented on CMakeLists.txt:
```diff
@@ -1088,19 +1116,74 @@ if(openPMD_BUILD_TESTING)
+                            --outfile ../samples/git-sample/thetaMode/data%T.bp;   \
+                                                                                   \
+                            ${Python_EXECUTABLE}                                   \
+                            ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/openpmd-pipe.py      \
```
```suggestion
                            ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/openpmd-pipe         \
```

ax3l commented on CMakeLists.txt:
```diff
@@ -1088,19 +1116,74 @@ if(openPMD_BUILD_TESTING)
+                add_test(NAME CLI.pipe.py
+                    COMMAND sh -c
+                        "${Python_EXECUTABLE}                                      \
+                            ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/openpmd-pipe.py      \
```
```suggestion
                            ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/openpmd-pipe         \
```

ax3l commented on CMakeLists.txt:
```diff
@@ -1088,19 +1116,74 @@ if(openPMD_BUILD_TESTING)
+                            --outfile ../samples/git-sample/thetaMode/data%T.bp;   \
+                                                                                   \
+                            ${Python_EXECUTABLE}                                   \
+                            ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/openpmd-pipe.py      \
```
```suggestion
                            ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/openpmd-pipe         \
```

ax3l commented on CMakeLists.txt:
```diff
@@ -1088,19 +1116,74 @@ if(openPMD_BUILD_TESTING)
+            test_set_pythonpath(CLI.help.${toolname}.py)
+        endforeach()
+
+        # openpmd-pipe.py test
```
```suggestion
        # openpmd-pipe (python) test
```

ax3l reviewed (dismissed):
.

ax3l reviewed (commented):
Great SFINAE-magic :D Hope it's not too wild on the compile time.

Overall this looks great :sparkles: 
I have some minor suggestions inline.

franzpoeschel commented:
> Hope it's not too wild on the compile time.

I'm planning a second PR after this one to apply those changes to the old `switchType` too. Since those functions don't really need to be in public headers, I would take the chance to move them to a separate header, that should have that issue solved.

franzpoeschel committed:
Make Attributable.attribute_dtypes a readonly property

franzpoeschel committed:
Introduce constants for HIGHEST/LOWEST_DATATYPE

franzpoeschel committed:
Rename Dummy -> Placeholder

franzpoeschel committed:
Don't suffix installed file with .py

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

ax3l reviewed (approved):
:rocket: :sparkles:
